pageserver/src/consumption_metrics.rs

```diff
@@ -222,38 +364,35 @@ pub async fn collect_metrics_iteration(
             }
         }
 
-        current_metrics.push((
-            PageserverConsumptionMetricsKey {
-                tenant_id,
-                timeline_id: None,
-                metric: RESIDENT_SIZE,
-            },
-            tenant_resident_size,
-        ));
+        current_metrics.push(
+            PageserverConsumptionMetricsKey::resident_size(tenant_id).now(tenant_resident_size),
+        );
 
         // Note that this metric is calculated in a separate bgworker
         // Here we only use cached value, which may lag behind the real latest one
         let tenant_synthetic_size = tenant.get_cached_synthetic_size();
 
         if tenant_synthetic_size != 0 {
             // only send non-zeroes because otherwise these show up as errors in logs
-            current_metrics.push((
-                PageserverConsumptionMetricsKey {
-                    tenant_id,
-                    timeline_id: None,
-                    metric: SYNTHETIC_STORAGE_SIZE,
-                },
-                tenant_synthetic_size,
-            ));
+            current_metrics.push(
+                PageserverConsumptionMetricsKey::synthetic_size(tenant_id)
+                    .now(tenant_synthetic_size),
+            );
         }
     }
 
     // Filter metrics, unless we want to send all metrics, including cached ones.
     // See: https://github.com/neondatabase/neon/issues/3485
     if !send_cached {
-        current_metrics.retain(|(curr_key, curr_val)| match cached_metrics.get(curr_key) {
-            Some(val) => val != curr_val,
-            None => true,
+        current_metrics.retain(|(curr_key, (_, curr_val))| {
+            if curr_key.is_written_size_delta() {
+                true
+            } else {
+                match cached_metrics.get(curr_key) {
+                    Some((_, val)) => val != curr_val,
+                    None => true,
+                }
+            }
         });
     }
```

Comment:
the is_written_size_delta is just a placeholder reminder.

Maybe we should never re-send incremental values? Unsure, it's very difficult to reason about this, given the idempotency keys are generated at the time of batching them up.

Reply:
Not as straightforward as I am using the abs time for last `last_record_lsn` now so that would have to be cached every time -- what is not retained or is removed from the metrics here are the new cached values. So if last_record_lsn would stop updating, we'd be pushing it with an ever growing time range.

Reply:
We will not cache the values which we fail to send, so in that regard this is working correctly. It would be nice to have the "refactored version" of this where we have separate actor for the http and retries, and per-tenant collection tasks...

Reply:
> We will not cache the values which we fail to send, so in that regard this is working correctly

Ah, so if prev values was V1, now is V2 and we fail to send delta `V2 - V1`, we will keep V1 as cached. Next time when we have V3, we will report `V3 - V1`? That's nice

Reply:
> Ah, so if prev values was V1, now is V2 and we fail to send delta V2 - V1, we will keep V1 as cached. Next time when we have V3, we will report V3 - V1? That's nice

Yes, exactly.

I'll push soon a change which will *even in the case* of having to send zero deltas give them the correct time range, which will be the latest `written_size_delta` to now. The value will be from the latest `written_size` minus previous `written_size` so the implementation is starting to look very ugly.